packages/components/src/tabs/tablist.tsx

```diff
@@ -17,19 +23,115 @@ import type { TabListProps } from './types';
 import { useTabsContext } from './context';
 import { TabListWrapper } from './styles';
 import type { WordPressComponentProps } from '../context';
+import type { CSSProperties } from 'react';
+import classNames from 'classnames';
+
+function useTrackElementOffset(
+	targetElement?: HTMLElement | null,
+	onUpdate?: () => void
+) {
+	const [ indicatorPosition, setIndicatorPosition ] = useState( {
+		left: 0,
+		top: 0,
+		width: 0,
+		height: 0,
+	} );
+
+	// TODO: replace with useEventCallback or similar when officially available.
+	const updateCallbackRef = useRef( onUpdate );
+	useLayoutEffect( () => {
+		updateCallbackRef.current = onUpdate;
+	} );
+
+	const observedElementRef = useRef< HTMLElement >();
+	const resizeObserverRef = useRef< ResizeObserver >();
+	useEffect( () => {
+		if ( targetElement === observedElementRef.current ) return;
+		observedElementRef.current = targetElement ?? undefined;
+
+		function updateIndicator( element: HTMLElement ) {
+			setIndicatorPosition( {
+				left: element.offsetLeft,
+				top: element.offsetTop,
+				width: element.offsetWidth,
+				height: element.offsetHeight,
+			} );
+			updateCallbackRef.current?.();
+		}
+
+		// Set up a ResizeObserver.
+		if ( ! resizeObserverRef.current ) {
+			resizeObserverRef.current = new ResizeObserver( () => {
+				if ( observedElementRef.current )
+					updateIndicator( observedElementRef.current );
+			} );
+		}
+		const { current: resizeObserver } = resizeObserverRef;
+
+		// Unobserve previous element.
+		const { current: observedElement } = observedElementRef;
+		if ( observedElement ) resizeObserver.unobserve( observedElement );
+
+		// Observe new element.
+		if ( targetElement ) {
+			updateIndicator( targetElement );
+			resizeObserver.observe( targetElement );
+		}
+	}, [ targetElement ] );
+
+	return indicatorPosition;
+}
+
+type ValueUpdateContext< T > = {
+	previousValue: T;
+};
+
+function useOnValueUpdate< T >(
+	value: T,
+	onUpdate: ( context: ValueUpdateContext< T > ) => void
+) {
+	const previousValueRef = useRef( value );
+
+	// TODO: replace with useEventCallback or similar when officially available.
+	const updateCallbackRef = useRef( onUpdate );
+	useLayoutEffect( () => {
+		updateCallbackRef.current = onUpdate;
+	} );
+
+	useEffect( () => {
+		if ( previousValueRef.current !== value ) {
+			updateCallbackRef.current( {
+				previousValue: previousValueRef.current,
+			} );
+			previousValueRef.current = value;
+		}
+	}, [ value ] );
+}
```

Comment:
Why do we need to maintain the update callback in a ref? What difference would it make if I did:

```suggestion
function useOnValueUpdate< T >(
	value: T,
	onUpdate: ( context: ValueUpdateContext< T > ) => void
) {
	const previousValueRef = useRef( value );

	useLayoutEffect( () => {
		if ( previousValueRef.current !== value ) {
			onUpdate( {
				previousValue: previousValueRef.current,
			} );
			previousValueRef.current = value;
		}
	}, [ value, onUpdate ] );
}
```

Reply:
This is a low-level event-oriented hook that needs to fire only on value updates. If the callback was a dependency for the effect, passing a different callback would trigger it, ergo the hook would not be firing on value updates exclusively. That would be misleading, and would also break the logic of consumers. In our case, such scenario would cause the animation to be enabled at a time when it shouldn't be.

Reply:
Btw, @jsnajdr also proposed this as I was about to push a commit with it, we had a parallel thinking moment :P https://github.com/WordPress/gutenberg/pull/60560#discussion_r1577505749